Updates button getter for new GamepadButton API

mildmojo opened the pull request:
Modern Firefox has aligned with the W3C Gamepad API spec, which calls for `gamepad.buttons` to be an array of GamepadButton objects rather than button state values. This adds support for that API. Tested in Firefox 29 for Linux.

I'm working on a patch to correctly map the PS4 Dualshock 4 controller, too, just need to make sure it behaves correctly on all platforms.

mildmojo committed:
Updates button getter for new GamepadButton API.

Modern Firefox has aligned with the W3C Gamepad API spec, which
calls for gamepad.buttons to be an array of GamepadButton objects
rather than button state values. This adds support for that API.

kallaspriit commented:
Have you checked other platforms and this is safe to merge?

mildmojo commented:
I haven't; I'll do that. In fact, it may fail on platforms that don't define GamepadButton. I'll probably need to check for `window.GamepadButton`, too.

cvan commented on gamepad.js:
```diff
@@ -836,7 +836,8 @@
 					getter = function() {
-						return gamepad.buttons[entry];
+						var button = gamepad.buttons[entry];
+						return button instanceof GamepadButton ? button.value : button;
```
you could add a check for `GamepadButton`:

``` js
'GamepadButton' in window && button instanceof GamepadButton
```

or you could change it to this:

``` js
typeof(button) === 'object' && button.value
```

cvan commented:
this library doesn't work for me without this patch. can we merge this in?

mildmojo commented:
I'm really sorry I haven't given this more attention; bunch of life changes and travel have kept me busy for a month or so. It's on me to test the patch more completely. I'll have a shot at testing these changes at a game dev meetup tonight. I'll try to incorporate @cvan's suggestion and test Firefox/Chrome/Chromium in Win/Mac/Linux.

mildmojo commented:
For some reason the fix for #23 in 2d39be9 is keeping the library from detecting my gamepad in Firefox 31 (Linux). Works before the commit, doesn't work after. Pressing face buttons on the `index.html` test page does nothing.

Looking at `Gamepad.resolvePlatform`, I'm not sure how it would ever select `FirefoxPlatform` since `isSupported()` just returns `true` for both the Firefox and WebKit platform classes and WebKit is the first one that `resolvePlatform` hits. 

Not sure if I should open this as a separate regression bug. I'll try to look at the code some more to understand how `resolvePlatform` is supposed to work.

cvan commented:
> For some reason the fix for #23 in 2d39be9 is keeping the library from detecting my gamepad in Firefox 31 (Linux). Works before the commit, doesn't work after. Pressing face buttons on the index.html test page does nothing.

That's because the unprefixed API uses `GamepadButton`s, which is what your pull addresses.

mildmojo commented:
No, this is happening before the pads are even detected. The return values for the gamepad getter functions in `WebKitPlatform.factory` are supposed to be gamepad arrays, but 2d39be9 introduced a bug where the getter functions themselves are returned instead. Here's the fix:

``` diff
                if (navigator) {
                        if (typeof(navigator.getGamepads) !== 'undefined') {
                                platform = new WebKitPlatform(listener, function() {
-                                       return navigator.getGamepads;
+                                       return navigator.getGamepads();
                                });
                        } else if (typeof(navigator.webkitGamepads) !== 'undefined') {
                                platform = new WebKitPlatform(listener, function() {
-                                       return navigator.webkitGamepads;
+                                       return navigator.webkitGamepads();
                                });
                        } else if (typeof(navigator.webkitGetGamepads) !== 'undefined') {
                                platform = new WebKitPlatform(listener, function() {
```

mildmojo committed:
Fixes regression in `WebKitPlatform.factory`.

Some `gamepadGetter` functions were returning the browser's gamepad
inquiry function instead of the gamepad list returned by that function.
Adds regression tests and fixes the issue.

mildmojo commented:
Tests are passing, including the tests I've added. Observed behavior w/r/t buttons and platforms:

_Linux Mint 17_:
- Firefox 31: OK
- Chromium ?: OK

_Windows 8:_
- Firefox 29?: All buttons detected as axes with `GamepadButton` objects as values.
- Chrome (latest): All buttons detected as axes with `GamepadButton` objects as values. 

_OSX 10.8.5:_
- Firefox 31: OK
- Chrome 36.0.1985.125: All buttons detected as axes with `GamepadButton` objects as values.

So that's odd. I need to dig deeper to figure out why my code is still allowing `GamepadButton` objects through as button values.

mildmojo commented:
I've tested that the current code works in:

_Linux Mint 17:_
- Firefox 31
- Chromium 34.0.1847.116

_OSX 10.8.5:_
- Firefox 31
- Chrome 36.0.1985.125

_Windows 8:_
- Firefox 30, 31
- Chrome 36.0.1985.125

Turns out there are a bunch of gotchas between platforms. Chrome for Windows doesn't expose the GamepadButton class used to construct the objects returned from the API, so you can't use `instanceof` tests. Firefox's button objects are some kind of subclass, so a duck-type check like `button.hasOwnProperty('value')` fails. :angry: 

Ready for review. :cactus:

kallaspriit commented:
Merged the pull request, thanks for the help :)
